### ISSUE TITLE
[v2] Typescript plugin not working

### ISSUE THREAD

**Issue by jaredpalmer** [gatsbyjs/gatsby#5520]
## Description

I tried to use the typescript plugin with v2 to no avail. I get the following error message.

```
ERROR in ./src/pages/index.tsx
Module parse failed: Unexpected token (3:7)
You may need an appropriate loader to handle this file type.
| import * as React from 'react';
|
| export interface IndexProps {}
|
| export interface IndexState {}
 @ ./.cache/sync-requires.js 11:51-124
 @ ./.cache/root.js
 @ ./.cache/app.js
 @ multi ./node_modules/react-hot-loader/patch.js (webpack)-hot-middleware/client.js?path=http://localhost:8000/__webpack_hmr&reload=true&overlay=false ./.cache/app
ℹ ｢wdm｣: Failed to compile.
```

### Environment
* Gatsby version (`npm list gatsby`): 2.0.0-alpha.40
* gatsby-cli version (`gatsby --version`):
* Node.js version: 8.9.x
* Operating System: macOS Sierra

### File contents (if changed)
`gatsby-config.js`: 
```js
module.exports = {
  plugins: [`gatsby-plugin-typescript`, `gatsby-plugin-react-helmet`],
};
```
`package.json`: N/A 
`gatsby-node.js`: N/A 
`gatsby-browser.js`: N/A 
`gatsby-ssr.js`: N/A

**Comment by KyleAMathews** [comment]
Hmm not sure — you can check out the source here to see if something looks awry https://github.com/gatsbyjs/gatsby/blob/v2/packages/gatsby-plugin-typescript/src/gatsby-node.js#L14

**Comment by jaredpalmer** [comment]
Why are you adding the js loaders here?

https://github.com/gatsbyjs/gatsby/blob/v2/packages/gatsby-plugin-typescript/src/gatsby-node.js#L39

**Comment by jaredpalmer** [comment]
hmmm `onCreateWebpack` config is not even getting fired... i just threw an error in there and nothin happened.
<img width="841" alt="screenshot 2018-05-22 16 32 45" src="https://user-images.githubusercontent.com/4060187/40388541-d4fce726-5ddd-11e8-8ed1-9ae3b863d713.png">

**Comment by KyleAMathews** [comment]
Hmm are any of the functions being called?

**Comment by jaredpalmer** [comment]
confirmed that resolvableExtensions is def getting called

**Comment by jaredpalmer** [comment]
confirmed that preprocessSource is too.

**Comment by KyleAMathews** [comment]
Odd — this is where the API is called — perhaps there's a subtle typo or something? https://github.com/gatsbyjs/gatsby/blob/81ebfc7661023f3c8e9c38a13f51d7055d0229bb/packages/gatsby/src/utils/webpack.config.js#L411

Is there any console errors?

**Comment by jaredpalmer** [comment]
oh well v2 of typescript is using `actions.setWebpackConfig` it's not merging and returning.

**Comment by jaredpalmer** [comment]
`actions` is not being passed to the code you just linked to.

**Comment by jquense** [comment]
> Why are you adding the js loaders here?

https://github.com/gatsbyjs/gatsby/pull/3369 

I've never felt like that was a good choice tho...especially now that babel has a TS parser

**Comment by mquandalle** [comment]
Indeed NextJS 6 uses babel parser for Typescript which integrates better with the build toolchain (you still get the TS experience in the editor)

**Comment by mquandalle** [comment]
I'm new to the Gatsby ecosystem and I have a Typescript app in which I would really benefit from webpack&nbsp;4 tree-shaking. So I'm trying the following configuration in `gatsby-node.js` with `gatsby@next`:

```es6
exports.resolvableExtensions = () => ['.ts', '.tsx']

exports.onCreateBabelConfig = ({ actions }, options) => {
  actions.setBabelPreset({
    name: '@babel/preset-typescript',
    options,
  })
}

exports.onCreateWebpackConfig = ({ actions, loaders }) => {
  actions.setWebpackConfig({
    module: {
      rules: [
        {
          test: /\.tsx?$/,
          use: [loaders.js()],
        },
      ],
    },
  })
}
```

Unfortunately this isn't working. Any pointer to what could go wrong here?

**Comment by KyleAMathews** [comment]
>Indeed NextJS 6 uses babel parser for Typescript which integrates better with the build toolchain 

Yeah, switching to a native babel parser sounds great (from my limited, haven't used typescript perspective)

**Comment by mquandalle** [comment]
@vadistic One possibility is to have type errors being reported in the code editor and not handled by the build toolchain (since Babel is only stripping TS type annotations). That's already the case with Gatsby v1 as your app compiles even if the TS type check fails.

Thanks for the pointer about `@babel/preset-flow`, I'll try to remove it and test the typescript preset instead.

**Comment by vadistic** [comment]
@jaredpalmer  Thanks for this! I was just spawning additional process for transpileOnly mode but it felt dirty and unsafe.

For now (you convinced (at least me) that babel+ ts-fork-checker would be way to go) - I've got current ts-loader plugin to work with Gatsby v2 (https://github.com/vadistic/gatsby-plugin-typescript).
1) Updated ts-loader to lattest
2) `delete typescriptOptions.plugins` from options passed to loader, because `ts-loader^4.x` has no such API
3) There's an error if you don't have tsconfig.json in a root from `ts-loader` but I left it unhandled - seems as desirable behavior for typescript project

Seems to work well, at least on some basic typescript starters - I cannot run the test suite and PR as unfortunately Gatsby multirepo fries my laptop 🚒

**Comment by dennari** [comment]
In case someone is looking for a temporary fix before `gatsby-plugin-typescript` is updated for v2, this seems to work:
1. Add `ts-loader` as dependency
2. Use the following `gatsby-node.js`:
```js
const { transpileModule } = require(`typescript`)

const test = /\.tsx?$/
const compilerDefaults = {
  target: `esnext`,
  experimentalDecorators: true,
  jsx: `react`,
  module: `es6`,
}

exports.resolvableExtensions = () => {
  return [`.ts`, `.tsx`]
}

exports.onCreateWebpackConfig = (
  { actions, loaders },
  { compilerOptions, ...options }
) => {
  const typescriptOptions = {
    transpileOnly: false,
    compilerOptions: {
      ...compilerDefaults,
      ...compilerOptions,
    },
  }
  actions.setWebpackConfig({
    module: {
      rules: [
        {
          test,
          use: [
            loaders.js(),
            {
              loader: require.resolve(`ts-loader`),
              options: typescriptOptions,
            },
          ],
        },
      ],
    },
  })
}

exports.preprocessSource = ({ contents, filename }, { compilerOptions }) => {
  const copts = { ...compilerDefaults, ...compilerOptions }

  // return the transpiled source if it's TypeScript, otherwise null
  return test.test(filename)
    ? transpileModule(contents, { compilerOptions: copts }).outputText
    : null
}
```
My `tsconfig.json` is:
```json
{
  "include": ["./src/**/*"],
  "compilerOptions": {
    "target": "esnext",
    "module": "commonjs",
    "lib": ["dom", "es2017"],
    "jsx": "react",
    "strict": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true
  }
}
```

**Comment by dennari** [comment]
Regarding the future of `gatsby-plugin-typescript`, I'd also vote for going the Babel route as was done with https://github.com/zeit/next-plugins/tree/master/packages/next-typescript. I'd be interested in contributing to this.

**Comment by dannywils** [comment]
I'm able to get a v2 project working with ts-loader with the guidance above but no static queries are running in dev or build. 

Debugging a bit leads me to believe the issue is related to this function: 

https://github.com/gatsbyjs/gatsby/blob/v2/packages/gatsby/src/internal-plugins/query-runner/file-parser.js#L106

The execution never seems to get to line 130. 

Anybody else running into this or have any ideas why typescript would affect static query execution?

**Comment by dennari** [comment]
@vadistic 

> gatsby loads general babel configs which contain preset `@babel/preset-flow`. This one is incompatible with `@babel/preset-typescript` and will throw errors together- you would need to remove flow preset and give it a try.

Looking at [this commit](https://github.com/babel/babel/commit/f2882d570ad656e4e99eb86930f7570e8f49d780#diff-b927af16ccd5c06377b32400f8eb774f), it seems that the two have been made compatible.

Btw, why is `@babel/preset-flow` included by default?

**Comment by jquense** [comment]
> Btw, why is @babel/preset-flow included by default?

My guess is that it was to reduce breakage in the babel7 migration, babel6's react preset included flow. Good question tho as to whether that is still a good idea..

**Comment by mquandalle** [comment]
I'm trying to use @dennari configuration, but I can't get it work. I'm under the impression that `preprocessSource` isn't called for `pages/` files. Could that be the case?

**Comment by mquandalle** [comment]
Hum actually I get the error

```
 — Your site's gatsby-node.js is exporting a variable named "onCreateWebpackConfig" which isn't an API. Perhaps you meant to export "onCreateBabelConfig"?
```

Is this API only available to plugins?

**Comment by dennari** [comment]
Got `preset-typescript` working (only tried with babel beta 49) with the following configuration:
```js
const resolvableExtensions = () => {
  return [`.ts`, `.tsx`]
}

const onCreateWebpackConfig = (
  { actions, loaders, stage },
  { compilerOptions, ...options }
) => {
  actions.setWebpackConfig({
    module: {
      rules: [
        {
          test: /\.tsx?$/,
          use: [
            {
              loader: require.resolve(`babel-loader`),
              options: {
                cacheDirectory: true,
                babelrc: false,
                sourceType: "unambiguous",
                presets: [
                  `@babel/preset-env`,
                  `@babel/preset-react`,
                  `@babel/preset-typescript`,
                ].map(require.resolve),
                plugins: [
                  require.resolve(`babel-plugin-remove-graphql-queries`),
                ],
              },
            },
          ],
        },
      ],
    },
  })
}
exports.onCreateWebpackConfig = onCreateWebpackConfig
exports.resolvableExtensions = resolvableExtensions
```
Here I'm not using `setBabelPreset`, since that seems to only affect `/.jsx?/` files.

This is only a quick fix though, since in a more general setup we should somehow be able to pass all the correct arguments to the presets and plugins. For comparison, here's a simplified version of the default rules for js/jsx parsing:

```js
{
  rules: [
    {
      test: /\.jsx?$/,
      exclude: /(node_modules|bower_components)/,
      use: [
        {
          options: {
            cacheDirectory: true,
            babelrc: false,
            sourceType: "unambiguous",
            presets: [
              [
                "preset-env",
                {
                  loose: true,
                  modules: false,
                  useBuiltIns: "usage",
                  targets: { browsers: [Array] },
                },
              ],
              [
                "preset-react",
                {
                  useBuiltIns: true,
                  pragma: "React.createElement",
                  development: false,
                },
              ],
              ["preset-flow", {}],
            ],
            plugins: [
              "babel-plugin-remove-graphql-queries",
              ["plugin-proposal-class-properties", { loose: true }],
              ["plugin-syntax-dynamic-import", {}],
              [
                "plugin-transform-runtime",
                { helpers: true, regenerator: true, polyfill: false },
              ],
            ],
          },
          loader: "babel-loader",
        },
      ],
    },
  ]
}

```

**Comment by dennari** [comment]
@dannywils , can you get the static query execution working with the above?

**Comment by dannywils** [comment]
It works but I had to force the preset to `@babel/preset-typescript 7.0.0-beta.42` because `7.0.0-beta.49 ` throws the following error:

```
Module build failed: Error: [BABEL] C:\dev\gatsby-typescript-staticquery\src\pages\index.tsx: .overrides is not allowed in preset options
    at Object.keys.forEach.key (C:\dev\gatsby-typescript-staticquery\node_modules\@babel\core\lib\config\validation\options.js:71:13)
```

Test repo at https://github.com/dannywils/gatsby-typescript-staticquery/tree/babel

**Comment by dennari** [comment]
I believe #5709 should solve the mentioned problems.

**Comment by mquandalle** [comment]
I'm using #5709 since a few days without issues. Also as noted before `7.0.0-beta.49` of `@babel/preset-typescript` was bugged, but the issue is now fixed in the latest `7.0.0-beta.51` so we don't have to pin to this particular beta version anymore.

**Comment by KyleAMathews** [comment]
Due to the high volume of issues, we're closing out older ones without recent activity. Please open a new issue if you need help!
